### PR TITLE
Server IO improvements and loading UX refresh

### LINKED ISSUE #440: [FR] Add option to toggle between per-field vs per-object-label coloring
Currently, colors in expanded sample view are the same for all detections in a given field. This is appropriate for comparing multiple sets of labels (e.g., ground truth vs predicted). However, in many situations, one only has a single label field. In this case, it would be desirable to have the option to show each object `label` in a different color, to help visually distinguish the different classes that the model is generating.

Likewise, for segmentation masks, it would be desirable to toggle between per-label-field coloring (where the segmentation mask is treated as a boolean `mask != 0` to determine which pixels to color in order to match the label field's color), or per-semantic-class coloring (which is what is currently supported in the App- each unique pixel value gets its own color).

This functionality is exposed via the core library when [drawing labels on samples](https://voxel51.com/docs/fiftyone/user_guide/draw_labels.html) as shown here:

```
annotation_config = foua.AnnotationConfig({
    "per_object_label_colors": True
})
```

A proposal is to add a `Per-object label colors` toggle button next to `Show/hide JSON` as shown below.

~~One may also want this option on the main grid view as well...~~

One will also want the ability to toggle these on the main grid view as well. See https://github.com/voxel51/fiftyone/pull/671 for a visual example.

root 2

### PR DESCRIPTION
**Pull request by benjaminpkane**
## What changes are proposed in this pull request?

To be reviewed after #668 is merged.

* Resolves #440 
* Replaces the Flask SocketIO web framework with Tornado
  * Pure websockets are now used, with asyncio concurrency
  * Multiple updates can be now pushed to clients (Apps or Sessions) without blocking each other
* Improved loading indicators
* Renames `Display Options` to `Fields`
* Introduces the `motor` mongodb client for asyncio based querying

TODOs:
 - [x] Test App updates from the view bar, and newly renamed `Display Options`
 - [x] Document `server/main.py`
 - [x] Clean up async `motor` usage in `core`

## How is this patch tested? If it is not, please explain why.

Locally. Server tests need to be updated.

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

See changes.

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

**Comment by benjaminpkane** [comment]
Still in progress. Will finish in the AM.

**Comment by benjaminpkane** [comment]
Undrafting. I still need to figure out to test the new WebSocket events on the server. Will hopefully resolve that soon.

Working through couple style tweaks, and some linting, as well. But this is ready for eyes.

**Comment by benjaminpkane** [comment]
I need to work on new web socket tests for the server, but I do not think it should block this PR. It will require some reading into `asyncio` and event loops.

This is a full change in web framework, so the figuring out how to best structure testing, and the server in general will be ongoing work.

Things are *much* faster now from my testing, though. Concurrency opens up a lot of doors in terms of handling larger datasets.

**Comment by brimoor** [comment]
Thanks for the color by label option!

How about putting the `refresh field colors` button inside the `options` menu so that it is hidden too when `options` is collapsed? Also, perhaps for visual consistency, you could make it look more like the other fields? Maybe this:

```
[gear] OPTIONS
---------------------------------
[  ] Color by label
<refresh icon> Refresh field colors
```

where clicking anywhere on the Refresh field colors row would cause the refresh action to happen.

Screen Shot 2020-11-12 at 2 21 01 PM

Screen Shot 2020-11-12 at 2 20 56 PM

**Comment by brimoor** [comment]
nit: can you add the vertical padding below the component that holds the selected label tags, like `(lane x)` below? It's currently flush with the bottom of the component

Screen Shot 2020-11-12 at 2 31 18 PM

**Comment by brimoor** [comment]
So I did the following:

```py
import fiftyone as fo
import fiftyone.zoo as foz

# Open a dataset that has a `uniqueness` field in the App
dataset, session = fo.quickstart()

# Now open a dataset with no `uniqueness` field in the App with the same session
session.dataset = foz.load_zoo_dataset("cifar10", split="test")
```

In the `Scalars` tab of the App, I curiously see a `uniqueness` graph even though the second dataset does not have a `uniqueness` field in its schema (only the first dataset did)

Screen Shot 2020-11-12 at 2 43 49 PM

**Review by brimoor — APPROVED** [review]
Wow this is a ton of work! 💪 💯 

I clicked around with the `quickstart`, `quickstart-video`, `bdd100k`, and `cifar10` datasets and things seemed to be working well.

I left a couple comments for your consideration, but I'm happy to approve now and let you handle the rest. Nice work!

File: fiftyone/core/session.py [hunk 1]
```diff
@@ -203,6 +203,8 @@ def dataset(self):
     @dataset.setter
     @_update_state
     def dataset(self, dataset):
+        if dataset is not None:
+            dataset._doc.reload()
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply fiftyone/core/session.py:207]: If I load a dataset, say the quickstart:
```py
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart")
```

And then look at the dataset `doc`, I get:
```
<DatasetDocument: {
    'id': '5fad964f2ef8900c501c6107',
    'media_type': None,
    'name': 'quickstart',
    'sample_collection_name': 'samples.2020.11.12.15.08.47',
    'persistent': False,
    'info': BaseDict({}),
    'sample_fields': EmbeddedDocumentList([
        <SampleFieldDocument: {
            'name': 'media_type',
            'ftype': 'fiftyone.core.fields.StringField',
            'subfield': None,
            'embedded_doc_type': None,
        }>,
        <SampleFieldDocument: {
            'name': 'filepath',
            'ftype': 'fiftyone.core.fields.StringField',
            'subfield': None,
            'embedded_doc_type': None,
        }>,
        <SampleFieldDocument: {
            'name': 'tags',
            'ftype': 'fiftyone.core.fields.ListField',
            'subfield': 'fiftyone.core.fields.StringField',
            'embedded_doc_type': None,
        }>,
        <SampleFieldDocument: {
            'name': 'metadata',
            'ftype': 'fiftyone.core.fields.EmbeddedDocumentField',
            'subfield': None,
            'embedded_doc_type': 'fiftyone.core.metadata.Metadata',
        }>,
        <SampleFieldDocument: {
            'name': '_rand',
            'ftype': 'fiftyone.core.fields.FloatField',
            'subfield': None,
            'embedded_doc_type': None,
        }>,
    ]),
    'frame_fields': EmbeddedDocumentList([]),
    'version': '0.6.5',
}>
```

I suppose this is a shortcoming of the `_serialize()` method on datasets. It pulls from the doc to get the schema. I want to circle back to it.

**Pull request merged**